### PR TITLE
backup: check the store state by last heartbeat

### LINKED ISSUE #42973: BR suffers a 15x performance regression when single TiKV node down
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. Create a cluster with 4 or more TiKV nodes.
2. Shut down one TiKV node.
3. Execute the backup.

### 2. What did you expect to see? (Required)
The backup should be slightly slower than backing up a healthy cluster.

### 3. What did you see instead (Required)
The backup speed is about 15x slower than backing up a healthy cluster. (4 mins vs 1 hour)
image

### 4. What is your TiDB version? (Required)
Near master, but this problem is not strong relative to the version.

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #42973

Problem Summary:
The current implementation cannot detect whether a store is down. So it keeps retry sending request to an offline store.

### What is changed and how it works?
This PR checks the `LastHeartbeatTS` of the store and give up to request the store once the gap is greater than 20s (also defined in the PD package.)

### Check List

Tests 

- [x] Manual test (add detailed scripts or steps below)
Backup over a cluster with one node down, finished in reasonable time.
image

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fixed a bug that caused backup get very slow when there are down TiKV nodes in the cluster.
```

### CODE REVIEW

File: br/pkg/backup/push.go [hunk 1]
```diff
@@ -51,6 +57,20 @@ func newPushDown(mgr ClientMgr, capacity int) *pushDown {
...
 
+func checkStoreLiveness(s *metapb.Store) error {
+	if s.State != metapb.StoreState_Up {
+		return errors.Annotatef(berrors.ErrKVStorage, "the store state isn't up, it is %s", s.State)
+	}
+	// If the field isn't present (the default value), skip this check.
+	if s.GetLastHeartbeat() > 0 {
+		lastHeartBeat := time.Unix(0, s.GetLastHeartbeat())
+		if sinceLastHB := time.Since(lastHeartBeat); sinceLastHB > storeDisconnectionDuration {
+			return errors.Annotatef(berrors.ErrKVStorage, "the store last heartbeat is too long ago, it is %s", sinceLastHB)
```
Comment (bufferflies) [inline br/pkg/backup/push.go:68]: how about logs the store id or more information?

File: br/pkg/backup/push.go [hunk 2]
```diff
@@ -22,6 +23,11 @@ import (
 	"go.uber.org/zap"
 )
 
+const (
+	// from PD: https://github.com/tikv/pd/blob/c40e319f50822678cda71ae62ee2fd70a9cac010/pkg/core/store.go#L523
+	storeDisconnectionDuration = 20 * time.Second
```
Comment (3pointer) [inline br/pkg/backup/push.go:28]: what if the value changed?

And can user change store heartbeat duration manually?

File: br/pkg/backup/push.go (hunk 1 above)
Comment (YuJuncen) [reply br/pkg/backup/push.go:68]: This information would be printed in the outside, like:
image

File: br/pkg/backup/push.go (hunk 2 above)
Comment (YuJuncen) [reply br/pkg/backup/push.go:28]: How can we configure this value? If you afraid this value is too pessimistic, perhaps we can make it longer. (say, 5 mins)

**Review by 3pointer — APPROVED** [review]

**Review by bufferflies — APPROVED** [review]

**Comment by hawkingrei** [comment]
> /test unit-test

I used to retest by /retest

**Comment by YuJuncen** [comment]
/retest

It seems we fail at different case each time...

**Pull request merged**